torch/utils/_sympy/symbol.py

```diff
@@ -76,20 +76,88 @@ class SymT(Enum):
 }
 
 
-def make_symbol(prefix: SymT, idx: int, **kwargs) -> sympy.Symbol:
-    # TODO: maybe put the assumptions here directly
-    return sympy.Symbol(f"{prefix_str[prefix]}{idx}", **kwargs)
+class PrefixedSymbol(sympy.Symbol):
+    def __new__(
+        cls, name, prefix, *, integer=None, nonnegative=None, positive=None, real=None
+    ):
+        obj = sympy.Symbol.__new__(
+            cls,
+            name,
+            integer=integer,
+            nonnegative=nonnegative,
+            positive=positive,
+            real=real,
+        )
+        obj.prefix = prefix
+        return obj
+
+    def _hashable_content(self):
+        # since we use only 3 assumptions, we only check on those three instead of the
+        # full set of assumptions to avoid an expensive caching operation
+        return (
+            self.name,
+            self.prefix,
+            self.is_integer,
+            self.is_nonnegative,
+            self.is_positive,
+            self.is_real,
+        )
+
+    def __copy__(self):
+        cls = self.__class__
+        return cls.__new__(
+            cls,
+            name=self.name,
+            prefix=self.prefix,
+            integer=self.is_integer,
+            nonnegative=self.is_nonnegative,
+            positive=self.is_positive,
+            real=self.is_real,
+        )
+
+    def __deepcopy__(self, memo):
+        result = self.__copy__()
+        memo[id(self)] = result
+        return result
+
+
+def make_symbol(
+    prefix: SymT, idx: int, integer=None, nonnegative=None, positive=None, real=None
+) -> sympy.Symbol:
+    return PrefixedSymbol(
+        f"{prefix_str[prefix]}{idx}",
+        prefix=prefix,
+        integer=integer,
+        nonnegative=nonnegative,
+        positive=positive,
+        real=real,
+    )
 
 
 # This type is a little wider than it should be, because free_symbols says
 # that it contains Basic, rather than Symbol
-def symbol_is_type(sym: sympy.Basic, prefix: Union[SymT, Sequence[SymT]]) -> bool:
-    assert isinstance(sym, sympy.Symbol)
-    name_str = sym.name.lower()  # Match capitalized names like XBLOCK, RBLOCK
-    if isinstance(prefix, SymT):
-        return name_str.startswith(prefix_str[prefix])
+def symbol_is_type(sym: sympy.Basic, prefix: Union[SymT, Tuple[SymT, ...]]) -> bool:
+    if isinstance(sym, PrefixedSymbol):
+        # This function is called a *lot* of times, so it needs to be fast
+        if type(prefix) == tuple:
+            # a list comprehension with any is slow
+            for p in prefix:
+                if sym.prefix == p:
+                    return True
+            return False
+        else:
+            return sym.prefix == prefix
     else:
-        return name_str.startswith(tuple(prefix_str[p] for p in prefix))
+        assert isinstance(sym, sympy.Symbol)
+        # TODO: remove all these Symbols and use PrefixedSymbol
+        sym_name = sym.name.lower()
```

Comment:
I hate this so frickin much lol. @shunting314, @jansel let's not take changes like this please lol (c.f. https://github.com/pytorch/pytorch/pull/127342 )